File: packages/android/src/main/kotlin/gov/civui/components/CivIcon.kt
// CivUI — CivIcon for Jetpack Compose
// Renders Material Symbols icons using the android mappings from icon-library.ts.
// Uses Material 3 Icon composable for consistent rendering and TalkBack support.

package gov.civui.components

import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.automirrored.filled.ArrowForward
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.filled.ArrowDownward
import androidx.compose.material.icons.filled.ArrowUpward
import androidx.compose.material.icons.filled.CalendarToday
import androidx.compose.material.icons.filled.Check
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material.icons.filled.ChevronLeft
import androidx.compose.material.icons.filled.ChevronRight
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.ContentCopy
import androidx.compose.material.icons.filled.Delete
import androidx.compose.material.icons.filled.DragHandle
import androidx.compose.material.icons.filled.Edit
import androidx.compose.material.icons.filled.Error
import androidx.compose.material.icons.filled.ExpandLess
import androidx.compose.material.icons.filled.ExpandMore
import androidx.compose.material.icons.filled.FilterList
import androidx.compose.material.icons.filled.FileDownload
import androidx.compose.material.icons.filled.FileUpload
import androidx.compose.material.icons.filled.Help
import androidx.compose.material.icons.filled.Home
import androidx.compose.material.icons.filled.Info
import androidx.compose.material.icons.filled.Lock
import androidx.compose.material.icons.filled.LocationOn
import androidx.compose.material.icons.filled.Mail
import androidx.compose.material.icons.filled.Menu
import androidx.compose.material.icons.filled.MoreHoriz
import androidx.compose.material.icons.filled.MoreVert
import androidx.compose.material.icons.filled.OpenInNew
import androidx.compose.material.icons.filled.Person
import androidx.compose.material.icons.filled.Print
import androidx.compose.material.icons.filled.Remove
import androidx.compose.material.icons.filled.Search
import androidx.compose.material.icons.filled.Settings
import androidx.compose.material.icons.filled.Star
import androidx.compose.material.icons.filled.StarBorder
import androidx.compose.material.icons.filled.Undo
import androidx.compose.material.icons.filled.UnfoldMore
import androidx.compose.material.icons.filled.Warning
import androidx.compose.material3.Icon
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import gov.civui.tokens.CivTokens

/**
 * Maps CivUI icon names (from icon-library.ts) to Material Icons vectors.
 * Keys match the `android` field in the web icon definitions.
 */
private val iconMap: Map<String, ImageVector> = mapOf(
    // Navigation
    "chevron_right" to Icons.Filled.ChevronRight,
    "chevron_left" to Icons.Filled.ChevronLeft,
    "expand_more" to Icons.Filled.ExpandMore,
    "expand_less" to Icons.Filled.ExpandLess,
    "arrow_forward" to Icons.AutoMirrored.Filled.ArrowForward,
    "arrow_back" to Icons.AutoMirrored.Filled.ArrowBack,
    "arrow_upward" to Icons.Filled.ArrowUpward,
    "arrow_downward" to Icons.Filled.ArrowDownward,
    "undo" to Icons.Filled.Undo,
    "open_in_new" to Icons.Filled.OpenInNew,

    // Actions
    "close" to Icons.Filled.Close,
    "add" to Icons.Filled.Add,
    "remove" to Icons.Filled.Remove,
    "menu" to Icons.Filled.Menu,
    "more_vert" to Icons.Filled.MoreVert,
    "more_horiz" to Icons.Filled.MoreHoriz,
    "search" to Icons.Filled.Search,
    "edit" to Icons.Filled.Edit,

    // Status
    "check" to Icons.Filled.Check,
    "check_circle" to Icons.Filled.CheckCircle,
    "error" to Icons.Filled.Error,
    "warning" to Icons.Filled.Warning,
    "info" to Icons.Filled.Info,
    "help" to Icons.Filled.Help,

    // Form
    "calendar_today" to Icons.Filled.CalendarToday,
    "location_on" to Icons.Filled.LocationOn,
    "unfold_more" to Icons.Filled.UnfoldMore,

    // Media
    "upload" to Icons.Filled.FileUpload,
    "download" to Icons.Filled.FileDownload,
    "filter_list" to Icons.Filled.FilterList,
    "content_copy" to Icons.Filled.ContentCopy,
    "delete" to Icons.Filled.Delete,

    // UI Chrome
    "drag_handle" to Icons.Filled.DragHandle,
    "lock" to Icons.Filled.Lock,
    "home" to Icons.Filled.Home,
    "settings" to Icons.Filled.Settings,
    "star_border" to Icons.Filled.StarBorder,
    "star" to Icons.Filled.Star,
    "print" to Icons.Filled.Print,
    "person" to Icons.Filled.Person,
    "mail" to Icons.Filled.Mail,
)

/**
 * Maps CivUI icon names (kebab-case, from web) to their Android material name.
 * Allows using either `"check-circle"` or `"check_circle"`.
 */
private val civNameToAndroid: Map<String, String> = mapOf(
    "chevron-right" to "chevron_right",
    "chevron-left" to "chevron_left",
    "chevron-down" to "expand_more",
    "chevron-up" to "expand_less",
    "arrow-right" to "arrow_forward",
    "arrow-left" to "arrow_back",
    "arrow-up" to "arrow_upward",
    "arrow-down" to "arrow_downward",
    "arrow-back" to "undo",
    "external-link" to "open_in_new",
    "close" to "close",
    "plus" to "add",
    "minus" to "remove",
    "menu" to "menu",
    "more-vertical" to "more_vert",
    "more-horizontal" to "more_horiz",
    "search" to "search",
    "edit" to "edit",
    "check" to "check",
    "check-circle" to "check_circle",
    "error" to "error",
    "warning" to "warning",
    "info" to "info",
    "help" to "help",
    "required-indicator" to "emergency",
    "sort-asc" to "arrow_upward",
    "sort-desc" to "arrow_downward",
    "sort-none" to "unfold_more",
    "calendar" to "calendar_today",
    "location" to "location_on",
    "upload" to "upload",
    "download" to "download",
    "filter" to "filter_list",
    "copy" to "content_copy",
    "trash" to "delete",
    "grip" to "drag_handle",
    "loading" to "progress_activity",
    "lock" to "lock",
    "home" to "home",
    "settings" to "settings",
    "star" to "star_border",
    "star-filled" to "star",
    "print" to "print",
    "user" to "person",
    "mail" to "mail",
)

/**
 * Resolves a CivUI icon name (kebab-case or Android underscore name)
 * to a Material Icons ImageVector.
 *
 * @return The ImageVector, or null if the icon name is not recognized.
 */
fun resolveIcon(name: String): ImageVector? {
    // Try direct Android name lookup first
    iconMap[name]?.let { return it }
    // Try CivUI kebab-case name
    val androidName = civNameToAndroid[name] ?: return null
    return iconMap[androidName]
}

/**
 * Accessible icon component for government applications.
 *
 * Renders a Material Symbol icon by CivUI name or Android name.
 * Supports decorative mode (no TalkBack announcement) and
 * informative mode (with content description).
 *
 * Usage:
 * ```kotlin
 * // Decorative icon (announced as part of surrounding text)
 * CivIcon(name = "check-circle")
 *
 * // Informative icon (TalkBack announces the label)
 * CivIcon(name = "warning", contentDescription = "Warning")
 *
 * // With custom color and size
 * CivIcon(
 *     name = "error",
 *     contentDescription = "Error",
 *     tint = CivTokens.Colors.Error.default_,
 *     size = 32.dp,
 * )
 * ```
 */
@Composable
fun CivIcon(
    name: String,
    modifier: Modifier = Modifier,
    contentDescription: String? = null,
    tint: Color = Color.Unspecified,
    size: Dp = 24.dp,
) {
    val isDark = isSystemInDarkTheme()
    val defaultTint = if (isDark) CivTokens.DarkColors.Base.darkest else CivTokens.Colors.Base.darkest

    val imageVector = resolveIcon(name) ?: return

    val effectiveTint = if (tint == Color.Unspecified) defaultTint else tint

    Icon(
        imageVector = imageVector,
        contentDescription = contentDescription,
        modifier = modifier.then(
            Modifier.padding(0.dp) // placeholder for size control
        ),
        tint = effectiveTint,
    )
}

File: packages/ios/Sources/CivUI/CivIcon.swift
// CivUI — CivIcon for SwiftUI
// Renders SF Symbols using the icon library mappings from @civui/core.
// Each icon name maps to an SF Symbol for native iOS rendering.

import SwiftUI

/// Icon name matching the web component's icon library.
/// Each case maps to an SF Symbol via the `ios` field in icon-library.ts.
public enum CivIconName: String, CaseIterable {
    // Navigation
    case chevronRight = "chevron-right"
    case chevronLeft = "chevron-left"
    case chevronDown = "chevron-down"
    case chevronUp = "chevron-up"
    case arrowRight = "arrow-right"
    case arrowLeft = "arrow-left"
    case arrowUp = "arrow-up"
    case arrowDown = "arrow-down"
    case arrowBack = "arrow-back"
    case externalLink = "external-link"

    // Actions
    case close
    case plus
    case minus
    case menu
    case moreVertical = "more-vertical"
    case moreHorizontal = "more-horizontal"
    case search
    case edit

    // Status / Feedback
    case check
    case checkCircle = "check-circle"
    case error
    case warning
    case info
    case help

    // Form / Input
    case requiredIndicator = "required-indicator"
    case sortAsc = "sort-asc"
    case sortDesc = "sort-desc"
    case sortNone = "sort-none"
    case calendar
    case location

    // Media / Content
    case upload
    case download
    case filter
    case copy
    case trash

    // UI Chrome
    case grip
    case loading
    case lock
    case home
    case settings
    case star
    case starFilled = "star-filled"
    case print
    case user
    case mail

    /// The SF Symbol name for this icon.
    public var sfSymbol: String {
        switch self {
        // Navigation
        case .chevronRight: return "chevron.right"
        case .chevronLeft: return "chevron.left"
        case .chevronDown: return "chevron.down"
        case .chevronUp: return "chevron.up"
        case .arrowRight: return "arrow.right"
        case .arrowLeft: return "arrow.left"
        case .arrowUp: return "arrow.up"
        case .arrowDown: return "arrow.down"
        case .arrowBack: return "arrow.uturn.backward"
        case .externalLink: return "arrow.up.right.square"

        // Actions
        case .close: return "xmark"
        case .plus: return "plus"
        case .minus: return "minus"
        case .menu: return "line.3.horizontal"
        case .moreVertical: return "ellipsis"
        case .moreHorizontal: return "ellipsis"
        case .search: return "magnifyingglass"
        case .edit: return "pencil"

        // Status
        case .check: return "checkmark"
        case .checkCircle: return "checkmark.circle"
        case .error: return "exclamationmark.circle"
        case .warning: return "exclamationmark.triangle"
        case .info: return "info.circle"
        case .help: return "questionmark.circle"

        // Form
        case .requiredIndicator: return "asterisk"
        case .sortAsc: return "chevron.up"
        case .sortDesc: return "chevron.down"
        case .sortNone: return "arrow.up.arrow.down"
        case .calendar: return "calendar"
        case .location: return "mappin.and.ellipse"

        // Media
        case .upload: return "arrow.up.doc"
        case .download: return "arrow.down.doc"
        case .filter: return "line.3.horizontal.decrease"
        case .copy: return "doc.on.doc"
        case .trash: return "trash"

        // UI Chrome
        case .grip: return "line.3.horizontal"
        case .loading: return "progress.indicator"
        case .lock: return "lock"
        case .home: return "house"
        case .settings: return "gearshape"
        case .star: return "star"
        case .starFilled: return "star.fill"
        case .print: return "printer"
        case .user: return "person"
        case .mail: return "envelope"
        }
    }

    /// Accessibility label for the icon.
    public var accessibilityLabel: String {
        switch self {
        case .chevronRight: return "Next"
        case .chevronLeft: return "Previous"
        case .chevronDown: return "Expand"
        case .chevronUp: return "Collapse"
        case .arrowRight: return "Right"
        case .arrowLeft: return "Left"
        case .arrowUp: return "Up"
        case .arrowDown: return "Down"
        case .arrowBack: return "Go back"
        case .externalLink: return "Opens in new tab"
        case .close: return "Close"
        case .plus: return "Add"
        case .minus: return "Remove"
        case .menu: return "Menu"
        case .moreVertical, .moreHorizontal: return "More options"
        case .search: return "Search"
        case .edit: return "Edit"
        case .check, .checkCircle: return "Success"
        case .error: return "Error"
        case .warning: return "Warning"
        case .info: return "Information"
        case .help: return "Help"
        case .requiredIndicator: return "Required"
        case .sortAsc: return "Sorted ascending"
        case .sortDesc: return "Sorted descending"
        case .sortNone: return "Unsorted"
        case .calendar: return "Calendar"
        case .location: return "Location"
        case .upload: return "Upload"
        case .download: return "Download"
        case .filter: return "Filter"
        case .copy: return "Copy"
        case .trash: return "Delete"
        case .grip: return "Drag handle"
        case .loading: return "Loading"
        case .lock: return "Locked"
        case .home: return "Home"
        case .settings: return "Settings"
        case .star: return "Favorite"
        case .starFilled: return "Favorited"
        case .print: return "Print"
        case .user: return "User account"
        case .mail: return "Email"
        }
    }
}

/// Accessible icon component using SF Symbols.
///
/// Maps CivUI icon names to their SF Symbol counterparts for native iOS rendering.
/// Supports decorative mode (hidden from VoiceOver) and custom sizing.
///
/// Usage:
/// ```swift
/// // Semantic icon (announced by VoiceOver)
/// CivIcon(.warning)
///
/// // Decorative icon (hidden from VoiceOver)
/// CivIcon(.chevronRight, decorative: true)
///
/// // Custom size and color
/// CivIcon(.checkCircle, size: 24)
///     .foregroundColor(CivTokens.Colors.Success.default_)
/// ```
public struct CivIcon: View {
    // MARK: - Properties

    /// The icon to display.
    public let name: CivIconName

    /// Icon size in points. Defaults to body text size.
    public var size: CGFloat

    /// Whether this icon is purely decorative (hidden from VoiceOver).
    public var decorative: Bool

    // MARK: - Initializer

    public init(
        _ name: CivIconName,
        size: CGFloat = CivTokens.Typography.FontSize.base,
        decorative: Bool = false
    ) {
        self.name = name
        self.size = size
        self.decorative = decorative
    }

    // MARK: - Body

    public var body: some View {
        Image(systemName: name.sfSymbol)
            .font(.system(size: size))
            .accessibilityLabel(decorative ? "" : name.accessibilityLabel)
            .accessibilityHidden(decorative)
    }
}

// MARK: - Preview

#if DEBUG
struct CivIcon_Previews: PreviewProvider {
    static var previews: some View {
        ScrollView {
            LazyVGrid(columns: [
                GridItem(.flexible()),
                GridItem(.flexible()),
                GridItem(.flexible()),
                GridItem(.flexible()),
            ], spacing: 16) {
                ForEach(CivIconName.allCases, id: \.self) { icon in
                    VStack(spacing: 4) {
                        CivIcon(icon, size: 24)
                        Text(icon.rawValue)
                            .font(.system(size: 10))
                            .lineLimit(1)
                    }
                    .frame(height: 60)
                }
            }
            .padding()
        }
        .previewDisplayName("Icon Gallery")
    }
}
#endif
